google/data_source_google_service_account_key.go

```diff
@@ -0,0 +1,88 @@
+package google
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func dataSourceGoogleServiceAccountKey() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceGoogleServiceAccountKeyRead,
+
+		Schema: map[string]*schema.Schema{
+			// Required
+			"service_account_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			// Optional
+			"public_key_type": &schema.Schema{
+				Type:         schema.TypeString,
+				Default:      "TYPE_X509_PEM_FILE",
+				Optional:     true,
+				ValidateFunc: validation.StringInSlice([]string{"TYPE_NONE", "TYPE_X509_PEM_FILE", "TYPE_RAW_PUBLIC_KEY"}, false),
+			},
+			"project": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			// Computed
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"key_algorithm": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"public_key": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceGoogleServiceAccountKeyRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	// Get the service account id as the fully qualified name
+	serviceAccountID := d.Get("service_account_id").(string)
+
+	// If the service account id isn't already the fully qualified name
+	if !strings.HasPrefix(serviceAccountID, "projects/") {
```

Comment:
How much of this can be shared?  Possibly even with the resources?

Reply:
Yup, I can share it with the resource `google_sevice_account_key`, but I don't know if I can do the same with `google_service_account` as it has `vadliateRFC1035Name(6, 30)` in the schema validation for `account_id` which doesn't allow '@' or `/`